### PR TITLE
[MODULAR] Size Collar

### PR DESCRIPTION
**Pull request by softcerv**
## About The Pull Request
Introduces a new item to the dorms vendor, the size collar. The size collar allows their user to change their size from as low as 15% up to 400%. However, _**the size changing function of the collar only works inside of dorms areas. It will revert the user to their normal size once they exit dorms.**_ Additionally, **before someone can use this item, they need to agree to a pop-up explaining that this item is not to be used outside of dorms or for any non-RP purpose.** 

Putting the collar on and adjusting the size changed by the collar are both logged. 

I needed to make a subtype of the interlink area for dorms, so that this wouldn't apply to the entire interlink.

## How This Contributes To The Skyrat Roleplay Experience
This allows for people that desire a certain kind of RP to preform that kind of RP. Also, because the effect of this item is entirely localized to dorms, this isn't something that is going to be seen outside of dorms.

## Proof of Testing

<details>
<summary>Screenshots/Videos</summary>
  
Warning message:
image

Worn inside of a dorms area:
image

Worn outside of a dorms area:
image

Max size:
image

Minimum size:
image

Appears in vendor:
image

Logging:
image

Interlink Dorm area:
image

</details>

## Changelog

:cl:
add: Adds in the size collar, an item obtainable from the dorm vendor that allows the user to change their size while inside of dorm rooms.
/:cl:

### CODE REVIEW

**Comment by tf-4** [comment]
I don't like the idea of magic "works only in dorms" items, conceptually

**Comment by softcerv** [comment]
> I don't like the idea of magic "works only in dorms" items, conceptually

This is the most likely the only way that something like this would be viable on our server. Unless you are fine with and willing to support the area lock on this being removed.

File: modular_skyrat/modules/modular_items/lewd_items/code/lewd_items/size_items.dm
Comment (tf-4) [inline modular_skyrat/modules/modular_items/lewd_items/code/lewd_items/size_items.dm]:
```suggestion
	log_message("[src] had its target size changed to [chosen_size]% by [usr]", LOG_ATTACK)
```

File: modular_skyrat/modules/modular_items/lewd_items/code/lewd_items/size_items.dm
Comment (tf-4) [inline modular_skyrat/modules/modular_items/lewd_items/code/lewd_items/size_items.dm]:
```suggestion
		balloon_alert(user, "invalid size!")
```

File: modular_skyrat/modules/modular_items/lewd_items/code/lewd_items/size_items.dm
Comment (tf-4) [inline modular_skyrat/modules/modular_items/lewd_items/code/lewd_items/size_items.dm]:
```suggestion
	balloon_alert(user, "set to [chosen_size]%")
```

File: modular_skyrat/modules/modular_items/lewd_items/code/lewd_items/size_items.dm
Comment (tf-4) [inline modular_skyrat/modules/modular_items/lewd_items/code/lewd_items/size_items.dm]:
```suggestion
	target_size = (chosen_size * 0.01)
```

File: modular_skyrat/modules/modular_items/lewd_items/code/lewd_items/size_items.dm
Comment (tf-4) [inline modular_skyrat/modules/modular_items/lewd_items/code/lewd_items/size_items.dm:10]: remember to clear references properly

File: modular_skyrat/modules/modular_items/lewd_items/code/lewd_items/size_items.dm
Comment (tf-4) [inline modular_skyrat/modules/modular_items/lewd_items/code/lewd_items/size_items.dm]:
```suggestion
	if(length(allowed_areas) && is_type_in_list(current_area, allowed_areas))
		apply_size(target_size)
		return TRUE

	apply_size(original_size)
	return FALSE
```
better to do it this way around, just in case

**Review by tf-4 — CHANGES_REQUESTED** [review]

**Comment by Elli-Skala** [comment]
Maybe instead of dorms it's effects fizzles out with a medium amount of tile movement so you can use it in non dorm hidey holes but not combatively

**Comment by Iamgoofball** [comment]
spaceman sprites get really, really ugly at a certain degree of resizing and i don't think this is a good idea even if it's magically locked to dorms

there's a certain point where you gotta draw the line and say "hey, use your imagination" when it comes to a text based roleplay medium, on top of the issues with ghosts having full vision meaning they're gonna see someone who's 600% xbox huge in a dorm room covering up the entire 3 adjacent dorm rooms

i'm having trouble envisioning a scenario where this is used by participants looking to do adult communication where they will not see the absurdity of a spaceman sprite that blown up and not immediately be taken out of the immersion

**Comment by softcerv** [comment]
> spaceman sprites get really, really ugly at a certain degree of resizing and i don't think this is a good idea even if it's magically locked to dorms because there's a certain point where you gotta draw the line and say "hey, use your imagination"

While I can understand your reasoning behind saying this, I think it's important to weigh the benefits vs drawbacks of this. 

> there's a certain point where you gotta draw the line and say "hey, use your imagination" when it comes to a text based roleplay medium, on top of the issues with ghosts having full vision meaning they're gonna see someone who's 600% xbox huge in a dorm room covering up the entire 3 adjacent dorm rooms

While yes, people can use this to create distorted sprites, said sprites are also entirely located within a very specific section of the station that most living players won't see. Regarding observers, I think that dorms are the least critical (and possibly most harmful) place for observers to be spectating. Nothing significant is lost if a dorm room is covered up. Although, I don't really see it being likely that enough players will be using the 600% mode for this to be a problem.

Meanwhile, this would be a massive boon for those that have certain "interests" that want to engage in a special kind of roleplay. Keep in mind that, without veteran crew, the maximum size a player is able to get their character, without using botany chemicals, is a sprite size of 150%. While this is big, some players would like to be bigger.

> i'm having trouble envisioning a scenario where this is used by participants looking to do adult communication where they will not see the absurdity of a spaceman sprite that blown up and not immediately be taken out of the immersion

Without trying to be rude, I think this may be the case because this is not something that is a personal "interest" of yours. What makes people feel immersed and un-immersed  in adult roleplay scenarios heavily ranges from person to person.

**Comment by Iamgoofball** [comment]
legit, fair

**Comment by SomeRandomOwl** [comment]
Being fully honest, I don't think this is necessary to add in. I agree with goof with some of their points, the sprites don't scale well when resized. Plus people can and already do just use their own imaginations when it comes to this sort of thing. When it boils down, these are pixels on the screen, half the actions people do are theater of the mind as a lot of emotes and such like that dont have animations or anything else of the nature. 

Regardless, I do not have final say in any of this. But my general opinion is this is trying to fix an issue that isnt a issue.

**Comment by softcerv** [comment]
> Being fully honest, I don't think this is necessary to add in. I agree with goof with some of their points, the sprites don't scale well when resized.

Only a minority of the playerbase is going to see the sprites in this state though.

> Plus people can and already do just use their own imaginations when it comes to this sort of thing. When it boils down, these are pixels on the screen, half the actions people do are theater of the mind as a lot of emotes and such like that dont have animations or anything else of the nature.

I'd ask you this then, why do we have any of the mechanics that we have in the first place when people can roleplay everything in the known universe out? The answer is flavor. Having bits and pieces of flavor can really help sell a roleplay scenerio, even if it's not completely necessary. This provides that flavor while having a negligible mechanical footprint. 

> Regardless, I do not have final say in any of this. But my general opinion is this is trying to fix an issue that isnt a issue.

I wouldn't say that I am trying to fix an issue here. I am more providing an additional feature to a set of people with a certain "interest" to help improve their roleplay.

**Comment by Gandalf2k15** [comment]
Given the already-stated feedback, I am not at all happy with this implementation. But, as a side note, I don't have an issue with fun things like this being accessible in VR(should someone ever implement this). But dorms is a publically accessible area and has no real protections as stated in the interlink change. Having massive or small sprites in dorms can and will be abused. But yes, see my alternative. It's a little OOS but hey, VR seems cool.

**Comment by UltimateFluff** [comment]
I'm sorry, this was closed because "I don't like this, use this thing I'm working on instead"? Gandalf, this honestly seems a bit...I don't know how to professionally word it. Especially since the only complaints about it are literately "I don't like this thing personally" and isn't even code-related. Honestly, if the only reason there are people voicing against it is because of personal preferences, then let me put it like this: Sex 2.0 was speed-merged in a broken state, adding a bunch of stuff that a number of people had no interest or liking of, and yet instead of it being removed and tinkered on to work better and relaunched? It stayed in, was worked on actively, and the people who couldn't deal with its existence left; those who didn't like it, but could deal with its existence? Just avoided it. Simple as that - because to pursue interacting with a thing you personally dislike constantly, then complain about this thing you dislike because you keep subjecting yourself to it? People are going to look at you, point, and laugh before telling you to just _not go fucking with it in the first place._

So if you aren't into this sort of thing personally? Word of advice: either keep your involvement professional as a coder, or just don't get yourself involved in the first place. Literately that easy.
Also, the fact that you just went ahead and closed it, Gandalf, because you want people to focus on **your** project instead of their own requested projects only makes it seem like overall interest in **your** project isn't where you want it to be, and you're wanting to force people to integrate with it. Maybe there's a logical, good reason for it - but from someone not involved in the code side of things, this feels malicious and predatory in nature, to close PRs and tell the people working on them to instead work on _your_ project and make it compatible with _it_ instead of doing what they were doing. Just doesn't look good, my guy.

**Comment by tf-4** [comment]
Literally what project are you talking about

**Comment by Iamgoofball** [comment]
```this feels malicious and predatory in nature```

could you elaborate on what is "predatory in nature" about a project maintainer closing a PR to the project they maintain? this seems like an odd critique of someone doing the job they've been assigned(make decisions on what goes in the codebase)

**Comment by Iamgoofball** [comment]
@softcerv if this does end up getting reopened I do recommend putting a note on the device to advise people use Nearest Neighbor scaling in the game options with this; otherwise it's going to be immensely blurry on the user's screen

i'm not sure if we default to nearest neighbor or not, i should check that

**Comment by softcerv** [comment]
> @softcerv if this does end up getting reopened I do recommend putting a note on the device to advise people use Nearest Neighbor scaling in the game options with this; otherwise it's going to be immensely blurry on the user's screen
> 
> i'm not sure if we default to nearest neighbor or not, i should check that

Honestly that's not a bad idea. Let me see if I can include it with the standard usage disclaimer.

**Comment by softcerv** [comment]
> Literally what project are you talking about

I think they are referring to Cobalt talking about limiting this to VR.

**Comment by tf-4** [comment]
That is not a project that currently exists

**Comment by UltimateFluff** [comment]
Okay, so let me tidy up things a bit on my end here.
1) Given that this isn't the first time the VR project has come up, and how they cited that looking at this PR being continued would be best focused on that, came across as being a case of "as a stand-alone, not ideal - rework it to go to something else that's being worked on instead." This leads to the next point, mind.
2) The VR project _still_ being a project not existing is, admittedly, a bit surprising - considering that a PR that's basically well on its way to being done already is being told to be shelved indefinitely for some _other_ project to get off the ground seems kind of...silly, I suppose?
3) The main issue people have apparently had with this PR, aside from the fact that they're not into it, is that they didn't read how it's **_literately restricted to the Interlink._** Which is why I said "if anyone bitches about it being a thing because they don't like it, they're purposely engaging in it because they want something to bitch about" - hence why I stated that, at least from a non-coder's perspective, all the complaints were worded as if from people who had no interest in it and why folks felt like they couldn't just...not be involved, or at least just keep it professional instead of saying "me no like, deny PR" with how they were coming across with their critique.

As far as I can tell so far, great care was taken to ensure that this requested feature was as low-impact as possible, and to get this much flak because the only people who chimed in the comments regarding it were people basically admitting that it wasn't their cup of tea, to me, seems a bit odd. Granted, I won't lie, I probably came off **way** too aggressively with it all, but it's a bit odd that it was closed, close to being essentially finished, on grounds of "the only comments left here were by people who don't like it" and nothing actually coding-related.
But again, I'm not a coder, so all I have to go on is what I see, and that's what I see.

**Pull request closed without merging**
